### PR TITLE
Fix precision issue with expansion that prefers 'probs' over 'logits'

### PR DESCRIPTION
**Pull request by ahmadsalim**
I have experienced that sometimes both were in `__dict__`, but it chose to copy `probs` which loses precision over `logits`. This is especially important when training (bayesian) neural networks or doing other type of optimization, since the loss is heavily affected.

### CODE REVIEW

File: torch/distributions/bernoulli.py [hunk 1]
```diff
@@ -50,12 +50,12 @@ def __init__(self, probs=None, logits=None, validate_args=None):
...
         new = self._get_checked_instance(Bernoulli, _instance)
         batch_shape = torch.Size(batch_shape)
-        if 'probs' in self.__dict__:
-            new.probs = self.probs.expand(batch_shape)
-            new._param = new.probs
-        else:
+        if 'logits' in self.__dict__:
             new.logits = self.logits.expand(batch_shape)
             new._param = new.logits
+        else:
```
Comment (neerajprad) [inline torch/distributions/bernoulli.py]: If both the parametrizations have already been computed, we should copy both of them, rather than having to recompute them again later, so I would suggest changing this to:

```python
if 'probs' in self.__dict__:
  ... # expand and set
if 'logits' in self.__dict__:
 ... # expand and set
```

That will also address your original issue since it will use the initially specified logits for any log pdf computation.

**Review by neerajprad — COMMENTED** [review]
Thanks for the PR! Could you also change this for the remaining distributions - binomial, multinomial, relaxed_bernoulli, negative_binomial, geometric?

**Comment by ahmadsalim** [comment]
Thanks for the comments! I will do the adjustments as you suggested for all the distributions 😄

**Comment by neerajprad** [comment]
LGTM! The lint check fails are not from this PR.

**Comment by soumith** [comment]
the CI is broken today, but I am putting it in the land queue.

**Review by soumith — APPROVED** [review]
@pytorchbot merge this please

**Review by neerajprad — APPROVED** [review]
